File: src/main/java/me/zeroeightsix/kami/module/modules/client/InventoryViewer.java
package me.zeroeightsix.kami.module.modules.client;

import me.zeroeightsix.kami.gui.rgui.component.container.use.Frame;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.ColourConverter;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;
import net.minecraft.util.ResourceLocation;
import org.lwjgl.opengl.GL11;

import static me.zeroeightsix.kami.util.GuiFrameUtil.getFrameByName;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendDisableMessage;

/**
 * Updated by dominikaaaa on 21/02/20
 * Slight updates by 20kdc, 19/02/20
 * Everything except somethingRender() methods was written by dominikaaaa
 */
@Module.Info(
        name = "InventoryViewer",
        category = Module.Category.CLIENT,
        description = "Configures Inventory Viewer's options",
        showOnArray = Module.ShowOnArray.OFF
)
public class InventoryViewer extends Module {
    private Setting<Boolean> mcTexture = register(Settings.b("UseResourcePack", false));
    private Setting<Boolean> showIcon = register(Settings.booleanBuilder("ShowIcon").withValue(true).withVisibility(v -> !mcTexture.getValue()).build());
    private Setting<ViewSize> viewSizeSetting = register(Settings.enumBuilder(ViewSize.class).withName("IconSize").withValue(ViewSize.LARGE).withVisibility(v -> showIcon.getValue() && !mcTexture.getValue()).build());
    private Setting<Boolean> coloredBackground = register(Settings.booleanBuilder("ColoredBackground").withValue(true).withVisibility(v -> !mcTexture.getValue()).build());
    private Setting<Integer> a = register(Settings.integerBuilder("Transparency").withMinimum(0).withValue(32).withMaximum(255).withVisibility(v -> coloredBackground.getValue() && !mcTexture.getValue()).build());
    private Setting<Integer> r = register(Settings.integerBuilder("Red").withMinimum(0).withValue(155).withMaximum(255).withVisibility(v -> coloredBackground.getValue() && !mcTexture.getValue()).build());
    private Setting<Integer> g = register(Settings.integerBuilder("Green").withMinimum(0).withValue(144).withMaximum(255).withVisibility(v -> coloredBackground.getValue() && !mcTexture.getValue()).build());
    private Setting<Integer> b = register(Settings.integerBuilder("Blue").withMinimum(0).withValue(255).withMaximum(255).withVisibility(v -> coloredBackground.getValue() && !mcTexture.getValue()).build());

    private ResourceLocation getBox() {
        if (mcTexture.getValue()) {
            return new ResourceLocation("textures/gui/container/generic_54.png");
        } else if (!showIcon.getValue()) {
            return new ResourceLocation("kamiblue/clear.png");
        } else if (viewSizeSetting.getValue().equals(ViewSize.LARGE)) {
            return new ResourceLocation("kamiblue/large.png");
        } else if (viewSizeSetting.getValue().equals(ViewSize.SMALL)) {
            return new ResourceLocation("kamiblue/small.png");
        } else if (viewSizeSetting.getValue().equals(ViewSize.MEDIUM)) {
            return new ResourceLocation("kamiblue/medium.png");
        } else {
            return new ResourceLocation("null");
        }
    }

    private enum ViewSize {
        LARGE, MEDIUM, SMALL
    }

    private void boxRender(final int x, final int y) {
        // SET UNRELIABLE DEFAULTS (Don't restore these) {
        GlStateManager.enableAlpha();
        GlStateManager.disableBlend();
        // }

        // ENABLE LOCAL CHANGES {
        GlStateManager.disableDepth();
        // }
        if (coloredBackground.getValue()) { // 1 == 2 px in game
            Gui.drawRect(x, y, x + 162, y + 54, ColourConverter.rgbToInt(r.getValue(), g.getValue(), b.getValue(), a.getValue()));
        }
        ResourceLocation box = getBox();
        mc.renderEngine.bindTexture(box);
        GlStateManager.color(1, 1, 1, 1);
        mc.ingameGUI.drawTexturedModalRect(x, y, 7, 17, 162, 54); // 164 56 // width and height of inventory
        // DISABLE LOCAL CHANGES {
        GlStateManager.enableDepth();
        // }
    }

    @Override
    public void onRender() {
        Frame frame = getFrameByName("inventory viewer");
        if (frame == null)
            return;
        if (frame.isPinned() && !frame.isMinimized()) {
            final NonNullList<ItemStack> items = mc.player.inventory.mainInventory;
            boxRender(frame.getX(), frame.getY());
            itemRender(items, frame.getX(), frame.getY());
        }
    }

    private void itemRender(final NonNullList<ItemStack> items, final int x, final int y) {
        GlStateManager.clear(GL11.GL_DEPTH_BUFFER_BIT);
        for (int size = items.size(), item = 9; item < size; ++item) {
            final int slotX = x + 1 + item % 9 * 18;
            final int slotY = y + 1 + (item / 9 - 1) * 18;
            //preItemRender(); Breaks with PlayerModel
            mc.getRenderItem().renderItemAndEffectIntoGUI(items.get(item), slotX, slotY);
            mc.getRenderItem().renderItemOverlays(mc.fontRenderer, items.get(item), slotX, slotY);
            //postItemRender(); Breaks with PlayerModel
        }
    }

    // These methods should apply and clean up in pairs.
    // That means that if a pre* has to disableAlpha, the post* function should enableAlpha.
    //  - 20kdc

    private static void preItemRender() {
        GlStateManager.pushMatrix();
        GlStateManager.enableDepth();
        GlStateManager.depthMask(true);
        // Yes, this is meant to be paired with disableStandardItemLighting - 20kdc
        RenderHelper.enableGUIStandardItemLighting();
    }

    private static void postItemRender() {
        RenderHelper.disableStandardItemLighting();
        GlStateManager.depthMask(false);
        GlStateManager.disableDepth();
        GlStateManager.popMatrix();
    }

    public void onDisable() { sendDisableMessage(this.getClass()); }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/client/InventoryViewer.kt
package me.zeroeightsix.kami.module.modules.client

import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Settings
import me.zeroeightsix.kami.util.color.ColorHolder
import me.zeroeightsix.kami.util.graphics.GlStateUtils
import me.zeroeightsix.kami.util.graphics.RenderUtils2D
import me.zeroeightsix.kami.util.graphics.VertexHelper
import me.zeroeightsix.kami.util.math.Vec2d
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.util.ResourceLocation
import org.lwjgl.opengl.GL11.*

@Module.Info(
        name = "InventoryViewer",
        category = Module.Category.CLIENT,
        description = "Configures Inventory Viewer's options",
        showOnArray = Module.ShowOnArray.OFF,
        alwaysEnabled = true
)
object InventoryViewer : Module() {
    private val mcTexture = register(Settings.b("UseMinecraftTexture", false))
    private val showIcon = register(Settings.booleanBuilder("ShowIcon").withValue(false).withVisibility { !mcTexture.value })
    private val iconScale = register(Settings.floatBuilder("IconScale").withValue(0.5f).withRange(0.1f, 1.0f).withStep(0.1f).withVisibility { !mcTexture.value && showIcon.value })
    private val coloredBackground = register(Settings.booleanBuilder("ColoredBackground").withValue(true).withVisibility { !mcTexture.value })
    private val r = register(Settings.integerBuilder("Red").withValue(155).withRange(0, 255).withStep(1).withVisibility { coloredBackground.value && !mcTexture.value })
    private val g = register(Settings.integerBuilder("Green").withValue(144).withRange(0, 255).withStep(1).withVisibility { coloredBackground.value && !mcTexture.value })
    private val b = register(Settings.integerBuilder("Blue").withValue(255).withRange(0, 255).withStep(1).withVisibility { coloredBackground.value && !mcTexture.value })
    private val a = register(Settings.integerBuilder("Alpha").withValue(32).withRange(0, 255).withStep(1).withVisibility { coloredBackground.value && !mcTexture.value })

    private val containerTexture = ResourceLocation("textures/gui/container/inventory.png")
    private val kamiIcon = ResourceLocation("kamiblue/kami_icon.png")

    fun renderInventoryViewer() {
        if (mc.player == null || mc.world == null) return
        drawFrame()
        drawFrameTexture()
        drawItems()
    }

    private fun drawFrame() {
        val vertexHelper = VertexHelper(GlStateUtils.useVbo())

        if (!mcTexture.value && coloredBackground.value) {
            RenderUtils2D.drawRectFilled(vertexHelper, posEnd = Vec2d(162.0, 54.0), color = ColorHolder(r.value, g.value, b.value, a.value))
        }
    }

    private fun drawFrameTexture() {
        val tessellator = Tessellator.getInstance()
        val buffer = tessellator.buffer
        GlStateUtils.texture2d(true)

        if (mcTexture.value) {
            mc.renderEngine.bindTexture(containerTexture)
            buffer.begin(GL_TRIANGLE_STRIP, DefaultVertexFormats.POSITION_TEX)
            buffer.pos(0.0, 0.0, 0.0).tex(0.02734375, 0.32421875).endVertex() // (1.75 / 64), (20.75 / 64)
            buffer.pos(0.0, 54.0, 0.0).tex(0.02734375, 0.53125).endVertex() // (1.75 / 64), (34 / 64)
            buffer.pos(162.0, 0.0, 0.0).tex(0.65625, 0.32421875).endVertex() // (42 / 64), (20.75 / 64)
            buffer.pos(162.0, 54.0, 0.0).tex(0.65625, 0.53125).endVertex() // (42 / 64), (34 / 64)
            tessellator.draw()
        } else if (showIcon.value) {
            mc.renderEngine.bindTexture(kamiIcon)
            GlStateManager.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)

            val center = Vec2d(81.0, 27.0)
            val halfWidth = iconScale.value * 54.0
            val halfHeight = iconScale.value * 27.0

            buffer.begin(GL_TRIANGLE_STRIP, DefaultVertexFormats.POSITION_TEX)
            buffer.pos(center.x - halfWidth, center.y - halfHeight, 0.0).tex(0.0, 0.0).endVertex()
            buffer.pos(center.x - halfWidth, center.y + halfHeight, 0.0).tex(0.0, 1.0).endVertex()
            buffer.pos(center.x + halfWidth, center.y - halfHeight, 0.0).tex(1.0, 0.0).endVertex()
            buffer.pos(center.x + halfWidth, center.y + halfHeight, 0.0).tex(1.0, 1.0).endVertex()
            tessellator.draw()

            GlStateManager.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)
        }
    }

    private fun drawItems() {
        val items = mc.player.inventory.mainInventory.subList(9, 36)

        for ((index, itemStack) in items.withIndex()) {
            val slotX = index % 9 * 18 + 1
            val slotY = index / 9 * 18 + 1

            GlStateUtils.blend(true)
            GlStateUtils.depth(true)
            RenderHelper.enableGUIStandardItemLighting()
            mc.renderItem.zLevel = 0.0f
            mc.renderItem.renderItemAndEffectIntoGUI(itemStack, slotX, slotY)
            mc.renderItem.renderItemOverlays(mc.fontRenderer, itemStack, slotX, slotY)
            mc.renderItem.zLevel = 0.0f
            RenderHelper.disableStandardItemLighting()
            GlStateManager.color(1.0f, 1.0f, 1.0f, 1.0f)
            GlStateUtils.depth(false)
        }
    }

}